graph/src/components/link_resolver.rs

```diff
@@ -1,71 +1,248 @@
 use crate::data::subgraph::Link;
+use bytes::BytesMut;
 use failure;
+use futures::{stream::poll_fn, try_ready};
 use ipfs_api;
+use serde_json::Value;
 use std::env;
 use std::str::FromStr;
 use std::time::Duration;
 use tokio::prelude::*;
 
 const MAX_IPFS_FILE_BYTES_ENV_VAR: &str = "GRAPH_MAX_IPFS_FILE_BYTES";
 
+const MAX_IPFS_MAP_FILE_SIZE_ENV_VAR: &str = "GRAPH_MAX_IPFS_MAP_FILE_SIZE";
+
+// Default size limitation for streaming files through ipfs_map is
+// 256MB
+const MAX_IPFS_MAP_FILE_SIZE: u64 = 256 * 1024 * 1024;
+
+type ValueStream = Box<Stream<Item = Value, Error = failure::Error> + Send + 'static>;
+
 /// Resolves links to subgraph manifests and resources referenced by them.
 pub trait LinkResolver: Send + Sync + 'static {
     /// Fetches the link contents as bytes.
     fn cat(&self, link: &Link) -> Box<Future<Item = Vec<u8>, Error = failure::Error> + Send>;
+
+    /// Read the contents of `link` and deserialize them into a stream of JSON
+    /// values. The values must each be on a single line; newlines are significant
+    /// as they are used to split the file contents and each line is deserialized
+    /// separately.
+    fn json_stream(
+        &self,
+        link: &Link,
+    ) -> Box<Future<Item = ValueStream, Error = failure::Error> + Send + 'static>;
+}
+
+// The timeout for IPFS requests in seconds
+fn ipfs_timeout() -> Duration {
+    let timeout = env::var("GRAPH_IPFS_TIMEOUT").ok().map(|s| {
+        u64::from_str(&s).unwrap_or_else(|_| panic!("failed to parse env var GRAPH_IPFS_TIMEOUT"))
+    });
+    Duration::from_secs(timeout.unwrap_or(30))
+}
+
+fn read_env_var(name: &str) -> Option<u64> {
+    env::var(name)
+        .ok()
+        .map(|s| u64::from_str(&s).unwrap_or_else(|_| panic!("failed to parse env var {}", name)))
```

Comment:
It may be worth mentioning `(expected: unsigned 64-bit integer)` or something.

Reply:
Good point, that was copy-pasta from the existing code. Changed the error message to `expected env var {} to contain a number (unsigned 64-bit integer), but got '{}'`